From dbbe4040299d2a02179cc390d9d7e2aab59833da Mon Sep 17 00:00:00 2001
From: Alexandre Derumier <aderumier@odiso.com>
Date: Mon, 22 May 2023 19:31:28 +0200
Subject: [PATCH] bgpd : add an option for RT auto-derivation to force AS

Allow to define a custom AS for autort AS:VNI

Usefull for evpn with ebgp, where we can't auto-derivated from the router AS,
as route-target AS:VNI need to be the same on the each peer

Link: https://github.com/FRRouting/frr/pull/7535
Signed-off-by: Alexandre Derumier <aderumier@odiso.com>
    [ GG: ported to latest stable 10.2.1 ]
Signed-off-by: Gabriel Goller <g.goller@proxmox.com>
---
 bgpd/bgp_evpn.c     | 13 +++++++++++--
 bgpd/bgp_evpn_vty.c | 46 +++++++++++++++++++++++++++++++++------------
 bgpd/bgpd.h         |  3 +++
 3 files changed, 48 insertions(+), 14 deletions(-)

Index: b/bgpd/bgp_evpn.c
===================================================================
--- a/bgpd/bgp_evpn.c	2025-03-07 11:09:47.315423809 +0100
+++ b/bgpd/bgp_evpn.c	2025-03-07 11:09:47.312423806 +0100
@@ -622,7 +622,12 @@
 
 	if (bgp->advertise_autort_rfc8365)
 		SET_FLAG(vni, EVPN_AUTORT_VXLAN);
-	encode_route_target_as((bgp->as & 0xFFFF), vni, &eval, true);
+
+	if (bgp->autort_as) {
+		encode_route_target_as((bgp->autort_as & 0xFFFF), vni, &eval, true);
+	}else {
+		encode_route_target_as((bgp->as & 0xFFFF), vni, &eval, true);
+	}
 
 	ecomadd = ecommunity_new();
 	ecommunity_add_val(ecomadd, &eval, false, false);
@@ -5490,7 +5495,11 @@
 	if (bgp->advertise_autort_rfc8365)
 		SET_FLAG(vni, EVPN_AUTORT_VXLAN);
 
-	encode_route_target_as((bgp->as & 0xFFFF), vni, &eval, true);
+	if (bgp->autort_as) {
+		encode_route_target_as((bgp->autort_as & 0xFFFF), vni, &eval, true);
+	}else {
+		encode_route_target_as((bgp->as & 0xFFFF), vni, &eval, true);
+	}
 
 	ecom_auto = ecommunity_new();
 	ecommunity_add_val(ecom_auto, &eval, false, false);
Index: b/bgpd/bgp_evpn_vty.c
===================================================================
--- a/bgpd/bgp_evpn_vty.c	2025-03-07 11:09:47.315423809 +0100
+++ b/bgpd/bgp_evpn_vty.c	2025-03-07 11:09:47.313423807 +0100
@@ -3744,32 +3744,51 @@
 	return CMD_SUCCESS;
 }
 
-DEFUN (bgp_evpn_advertise_autort_rfc8365,
-       bgp_evpn_advertise_autort_rfc8365_cmd,
-       "autort rfc8365-compatible",
+DEFPY (bgp_evpn_advertise_autort,
+       bgp_evpn_advertise_autort_cmd,
+       "autort [rfc8365-compatible]$rfc8365 [as (1-65536)$as]",
+       NO_STR
        "Auto-derivation of RT\n"
-       "Auto-derivation of RT using RFC8365\n")
+       "Auto-derivation of RT using RFC8365\n"
+       "Auto-derivation AS\n")
 {
 	struct bgp *bgp = VTY_GET_CONTEXT(bgp);
 
 	if (!bgp)
 		return CMD_WARNING;
-	evpn_set_advertise_autort_rfc8365(bgp);
+
+	if (as) {
+		bgp->autort_as = as;
+		bgp_evpn_handle_autort_change(bgp);
+	}
+
+	if (rfc8365)
+		evpn_set_advertise_autort_rfc8365(bgp);
+
 	return CMD_SUCCESS;
 }
 
-DEFUN (no_bgp_evpn_advertise_autort_rfc8365,
-       no_bgp_evpn_advertise_autort_rfc8365_cmd,
-       "no autort rfc8365-compatible",
+DEFPY (no_bgp_evpn_advertise_autort,
+       no_bgp_evpn_advertise_autort_cmd,
+       "no autort [rfc8365-compatible]$rfc8365 [as (1-65536)$as]",
        NO_STR
        "Auto-derivation of RT\n"
-       "Auto-derivation of RT using RFC8365\n")
+       "Auto-derivation of RT using RFC8365\n"
+       "Auto-derivation AS\n")
 {
 	struct bgp *bgp = VTY_GET_CONTEXT(bgp);
 
 	if (!bgp)
 		return CMD_WARNING;
-	evpn_unset_advertise_autort_rfc8365(bgp);
+
+	if (as) {
+		bgp->autort_as = 0;
+		bgp_evpn_handle_autort_change(bgp);
+	}
+
+	if (rfc8365)
+		evpn_unset_advertise_autort_rfc8365(bgp);
+
 	return CMD_SUCCESS;
 }
 
@@ -7286,6 +7305,9 @@
 	if (bgp->advertise_autort_rfc8365)
 		vty_out(vty, "  autort rfc8365-compatible\n");
 
+	if (bgp->autort_as)
+		vty_out(vty, "  autort as %u\n", bgp->autort_as);
+
 	if (bgp->advertise_gw_macip)
 		vty_out(vty, "  advertise-default-gw\n");
 
@@ -7528,8 +7550,8 @@
 	install_element(BGP_EVPN_NODE, &evpnrt5_network_cmd);
 	install_element(BGP_EVPN_NODE, &bgp_evpn_advertise_all_vni_cmd);
 	install_element(BGP_EVPN_NODE, &no_bgp_evpn_advertise_all_vni_cmd);
-	install_element(BGP_EVPN_NODE, &bgp_evpn_advertise_autort_rfc8365_cmd);
-	install_element(BGP_EVPN_NODE, &no_bgp_evpn_advertise_autort_rfc8365_cmd);
+	install_element(BGP_EVPN_NODE, &bgp_evpn_advertise_autort_cmd);
+	install_element(BGP_EVPN_NODE, &no_bgp_evpn_advertise_autort_cmd);
 	install_element(BGP_EVPN_NODE, &bgp_evpn_advertise_default_gw_cmd);
 	install_element(BGP_EVPN_NODE, &no_bgp_evpn_advertise_default_gw_cmd);
 	install_element(BGP_EVPN_NODE, &bgp_evpn_advertise_svi_ip_cmd);
Index: b/bgpd/bgpd.h
===================================================================
--- a/bgpd/bgpd.h	2025-03-07 11:09:47.315423809 +0100
+++ b/bgpd/bgpd.h	2025-03-07 11:09:47.313423807 +0100
@@ -745,6 +745,9 @@
 	/* EVPN - use RFC 8365 to auto-derive RT */
 	int advertise_autort_rfc8365;
 
+	/* EVPN - auto-derive RT AS */
+	uint16_t autort_as;
+
 	/*
 	 * Flooding mechanism for BUM packets for VxLAN-EVPN.
 	 */

